From c775411174d0605903e4270a4f1654b388cd6063 Mon Sep 17 00:00:00 2001
From: Emergi Idan <idan.emergi@intel.com>
Date: Tue, 29 Sep 2020 13:52:01 +0300
Subject: [PATCH] LMTSYS-2485: Fix PPv4 QoS FW elf size calculation

Allocate memory based on elf last section address field instead
of the total sizes of all sections
---
 drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c
index 6ec7a77d041d..dc0a2f1bb817 100644
--- a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c
+++ b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c
@@ -107,8 +107,13 @@ static void calc_elf_sections_sizes(struct elf_sec secs[], u16 num_secs,
 	u16 ind;
 
 	for (ind = 0; ind < num_secs; ind++) {
+		/* the addr field isn't in sequence after the previous one,
+		 * and we must keep all sections in same address(offset)
+		 * as defined by the elf file, so the total size is where
+		 * the last section that should be copied ends.
+		 */
 		if (secs[ind].need_copy)
-			*total_sz += secs[ind].size;
+			*total_sz = secs[ind].addr + secs[ind].size;
 
 		if (!strncmp(secs[ind].name, ".data", 5))
 			*data_sz = secs[ind].size;
